File: api/tts-with-timing.py
"""
Vercel Python Serverless Function: POST /api/tts-with-timing
Body (JSON): { text, voice, rate="+0%", pitch="+0Hz" }
Response: JSON { audio: number[], boundaries: [], duration: 0 }
"""
import json
import asyncio
import traceback

try:
    import edge_tts
except ImportError as e:
    def handler(event, context):
        return {
            "statusCode": 500,
            "headers": {
                "Content-Type": "application/json",
                "Access-Control-Allow-Origin": "*",
                "Access-Control-Allow-Headers": "Content-Type",
                "Access-Control-Allow-Methods": "POST,OPTIONS",
            },
            "body": json.dumps({"error": f"edge-tts not installed ({e}). Add requirements.txt with edge-tts"}),
        }


async def _synthesize(text, voice, rate, pitch):
    communicate = edge_tts.Communicate(text, voice, rate=rate, pitch=pitch)
    audio_bytes = bytearray()
    boundaries = []
    async for chunk in communicate.stream():
        t = chunk.get("type")
        if t == "audio":
            audio_bytes.extend(chunk["data"])
        elif t == "WordBoundary":
            boundaries.append({
                "offset": chunk.get("offset", 0),
                "duration": chunk.get("duration", 0),
                "text": chunk.get("text", ""),
            })
    return bytes(audio_bytes), boundaries


def handler(event, context):
    method = event.get("httpMethod") or event.get("method", "GET")
    headers = {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "Content-Type",
        "Access-Control-Allow-Methods": "POST,OPTIONS",
    }
    if method == "OPTIONS":
        return {"statusCode": 200, "headers": headers, "body": ""}
    if method != "POST":
        return {"statusCode": 405, "headers": headers, "body": json.dumps({"error": "Method Not Allowed"})}
    try:
        body_raw = event.get("body", "{}")
        if isinstance(body_raw, (bytes, bytearray)):
            body_raw = body_raw.decode("utf-8")
        if event.get("isBase64Encoded") and body_raw:
            import base64 as _b64
            body_raw = _b64.b64decode(body_raw).decode("utf-8")
        payload = json.loads(body_raw or "{}")
    except Exception as e:
        return {"statusCode": 400, "headers": headers, "body": json.dumps({"error": f"Invalid JSON: {e}"})}

    text = str(payload.get("text", "")).strip()
    voice = str(payload.get("voice", "")).strip()
    rate = str(payload.get("rate", "+0%")).strip() or "+0%"
    pitch = str(payload.get("pitch", "+0Hz")).strip() or "+0Hz"

    if not text:
        return {"statusCode": 400, "headers": headers, "body": json.dumps({"error": "Text required"})}
    if not voice:
        return {"statusCode": 400, "headers": headers, "body": json.dumps({"error": "Voice required"})}
    if len(text) > 2500:
        return {"statusCode": 413, "headers": headers, "body": json.dumps({"error": "Text > 2500 chars"})}

    try:
        audio_bytes, boundaries = asyncio.run(_synthesize(text, voice, rate, pitch))
        if not audio_bytes or len(audio_bytes) < 100:
            return {"statusCode": 500, "headers": headers, "body": json.dumps({"error": "Audio buffer quá nhỏ"})}
        duration = 0
        if boundaries:
            last = boundaries[-1]
            duration = (last.get("offset", 0) + last.get("duration", 0)) / 10_000_000.0
        return {
            "statusCode": 200,
            "headers": headers,
            "body": json.dumps({
                "audio": list(audio_bytes),
                "boundaries": boundaries,
                "duration": duration,
            }),
        }
    except Exception as e:
        tb = traceback.format_exc()
        return {"statusCode": 500, "headers": headers, "body": json.dumps({"error": f"{e}", "trace": tb[:500]})}

File: api/tts-with-timing.js
import { EdgeTTS } from 'node-edge-tts';
import fs from 'fs';
import os from 'os';
import path from 'path';

export default async function handler(req, res) {
  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method Not Allowed' });
  }

  const { text, voice, rate = '+0%', pitch = '+0Hz' } = req.body || {};
  if (!text) return res.status(400).json({ error: 'Text is required' });
  if (!voice) return res.status(400).json({ error: 'Voice is required' });
  if (text.length > 2500) return res.status(413).json({ error: 'Text quá dài (>2500 ký tự).' });

  // Timeout: Vercel Hobby kills at 10s, respond before that
  const timeout = setTimeout(() => {
    if (!res.writableEnded) {
      res.status(504).json({ error: 'Request timeout (>9s). Vui lòng thử lại.', trace: '' });
    }
  }, 9000);

  try {
    const result = await _synthesize(text, voice, rate, pitch);
    clearTimeout(timeout);
    if (!res.writableEnded) {
      const b64 = result.audio.toString('base64');
      return res.status(200).json({ audio: b64, words: result.words });
    }
  } catch (err) {
    clearTimeout(timeout);
    if (!res.writableEnded) {
      return res.status(500).json({
        error: err.message || 'Internal server error',
        trace: (err.stack || '').slice(0, 2000),
      });
    }
  }
}

async function _synthesize(text, voice, rate, pitch) {
  for (let attempt = 0; attempt < 3; attempt++) {
    const tmpFile = path.join(os.tmpdir(), `tts_${Date.now()}_${attempt}.mp3`);
    const subFile = tmpFile + '.json';
    try {
      const tts = new EdgeTTS({
        voice,
        rate,
        pitch,
        volume: '+0%',
        lang: voice.startsWith('vi') ? 'vi-VN' : voice.slice(0, 5),
        saveSubtitles: true,
        timeout: 8000,
      });
      await tts.ttsPromise(text, tmpFile);
      const audioBuffer = fs.readFileSync(tmpFile);
      if (!audioBuffer || audioBuffer.length < 100) {
        if (attempt < 2) {
          await new Promise(r => setTimeout(r, 1000 * (attempt + 1)));
          continue;
        }
        throw new Error('Không tạo được âm thanh (buffer quá nhỏ).');
      }
      let words = [];
      try {
        const subData = fs.readFileSync(subFile, 'utf-8');
        words = JSON.parse(subData);
      } catch { /* subtitles file may not exist */ }
      return { audio: audioBuffer, words };
    } catch (err) {
      if (attempt < 2 && err.message?.includes('No audio')) {
        await new Promise(r => setTimeout(r, 1000 * (attempt + 1)));
        continue;
      }
      throw err;
    } finally {
      try { fs.unlinkSync(tmpFile); } catch { /* ignore */ }
      try { fs.unlinkSync(subFile); } catch { /* ignore */ }
    }
  }
}